googletest/src/gtest.cc

```diff
@@ -723,60 +723,92 @@ static bool PatternMatchesString(const std::string& name_str,
   return true;
 }
 
-bool UnitTestOptions::MatchesFilter(const std::string& name_str,
-                                    const char* filter) {
-  // The filter is a list of patterns separated by colons (:).
-  const char* pattern = filter;
-  while (true) {
-    // Find the bounds of this pattern.
-    const char* const next_sep = strchr(pattern, ':');
-    const char* const pattern_end =
-        next_sep != nullptr ? next_sep : pattern + strlen(pattern);
-
-    // Check if this pattern matches name_str.
-    if (PatternMatchesString(name_str, pattern, pattern_end)) {
-      return true;
-    }
+namespace {
+class UnitTestFilter {
+ public:
+  UnitTestFilter() = default;
+
+  // Constructs a filter form a string of patterns separated by `:`.
+  explicit UnitTestFilter(const std::string& filter) {
+    if (filter.empty()) return;
+
+    SplitString(filter, ':', &patterns_);
+  }
 
-    // Give up on this pattern. However, if we found a pattern separator (:),
-    // advance to the next pattern (skipping over the separator) and restart.
-    if (next_sep == nullptr) {
-      return false;
+  // Returns true if and only if name matches at least one of the patterns in
+  // the filter.
+  bool MatchesName(const std::string& name) const {
+    const auto pattern_matches_name = [&name](const std::string& pattern) {
+      return PatternMatchesString(name, pattern.c_str(),
+                                  pattern.c_str() + pattern.size());
+    };
+    return std::any_of(patterns_.begin(), patterns_.end(),
+                       pattern_matches_name);
+  }
+
+ private:
+  std::vector<std::string> patterns_;
+};
+
+class PositiveAndNegativeUnitTestFilter {
+ public:
+  // Constructs a positive and a negative filter from a string. The string
+  // contains a positive filter optionally followed by a '-' character and a
+  // negative filter. In case only a negative filter is provided the positive
+  // filter will be assumed "*".
+  // A filter is a list of patterns separated by ':'.
+  explicit PositiveAndNegativeUnitTestFilter(const std::string& filter) {
+    std::vector<std::string> positive_and_negative_filters;
+
+    SplitString(filter, '-', &positive_and_negative_filters);
+    const auto& positive_filter = positive_and_negative_filters.front();
+
+    if (positive_and_negative_filters.size() > 1) {
```

Comment:
Unfortunately, the original API did not do any validation here -- instead, it just split on the first `-` and ignored others. This would change it to split on all `-` and only consider the first and the last segments if there were more than 2.

Even though it's undefined behavior, I'd retain the original to avoid breaking [inevitable](https://www.hyrumslaw.com/) dependents for now, e.g. by concatenating everything past the first element into the negative filter.
Since we're here, please add a comment like `// TODO: Fail on multiple '-' characters`.

Reply:
Sure this is the only behavior change this merge request contains, AFAIK. I will fix this, and hopefully the tests pass.